### DISCUSSION TITLE
Fps

### DISCUSSION THREAD

**Discussion by Eaielectronic** [TurboWarp/extensions#1711] [Ideas]
Hello, it would be nice if there was an option to change only the FPS, one to only change the execution time of the sprites and one for both at the same time.

**Comment by Thebloxers998** [comment]
Nice idea!

**Comment by SharkPool-SP** [comment]
> Hello, it would be nice if there was an option to change only the FPS, one to only change the execution time of the sprites and one for both at the same time.

runtime extension has a set fps block

**Reply by Thebloxers998** [reply]
Oh yea

**Comment by Eaielectronic** [comment]
I created this extension to keep the same script execution time. A block that returns 30 times 1 every 1 second :`lass CombinedExtension {
    constructor() {
        this.previousTime = null;  // Stocke l'heure de la dernière frame
        this.count = 0;            // Compte le nombre de fois où on a retourné 1
    }

    getInfo() {
        return {
            id: 'combinedextension',
            name: 'Combined Delta Time & Return 1',
            blocks: [
                {
                    opcode: 'getDeltaTime',
                    blockType: Scratch.BlockType.REPORTER,
                    text: 'delta time'
                },
                {
                    opcode: 'getFixedDelta',
                    blockType: Scratch.BlockType.REPORTER,
                    text: 'fixed delta (target 1/sec)'
                },
                {
                    opcode: 'returnOneMultipleTimes',
                    blockType: Scratch.BlockType.REPORTER,
                    text: 'return 30 times 1 every second'
                },
                {
                    opcode: 'getIncrementForDuration',
                    blockType: Scratch.BlockType.REPORTER,
                    text: 'increment to reach [TARGET] in [DURATION] sec',
                    arguments: {
                        TARGET: {
                            type: Scratch.ArgumentType.NUMBER,
                            defaultValue: 1
                        },
                        DURATION: {
                            type: Scratch.ArgumentType.NUMBER,
                            defaultValue: 3
                        }
                    }
                }
            ]
        };
    }

    // Méthode pour obtenir le delta time
    getDeltaTime() {
        const currentTime = performance.now();

        if (this.previousTime === null) {
            this.previousTime = currentTime;
            return 0;
        }

        let deltaTime = (currentTime - this.previousTime) / 1000; // Convertir en secondes
        this.previousTime = currentTime;

        if (deltaTime > 1/30) {
            deltaTime = 1/30;  // Limiter à 30 FPS max pour éviter les pics
        }

        return deltaTime;
    }

    // Méthode pour obtenir un delta fixe
    getFixedDelta() {
        const deltaTime = this.getDeltaTime();
        return deltaTime * 100;  // Calibrer pour que "100 * delta time" = 1 unité ajoutée par seconde
    }

    // Méthode pour obtenir un incrément en fonction de la durée pour atteindre une cible
    getIncrementForDuration(args) {
        const deltaTime = this.getDeltaTime();
        const targetValue = args.TARGET;   // Valeur cible que vous voulez atteindre
        const duration = args.DURATION;    // Durée en secondes pour atteindre la valeur

        // Chaque seconde, on ajoute une fraction pour atteindre la valeur cible dans le temps imparti
        return (targetValue / duration) * deltaTime;
    }

    // Méthode pour retourner 1 plusieurs fois par seconde
    returnOneMultipleTimes() {
        const deltaTime = this.getDeltaTime();
        const returnsPerSecond = 30;  // Retourner 30 fois 1 par seconde
        const returns = deltaTime * returnsPerSecond;  // Combien de fois retourner 1 durant deltaTime

        this.count += returns;

        if (this.count >= 1) {
            this.count -= 1;
            return 1;
        } else {
            return 0;  // Retourner 0 si le temps écoulé ne justifie pas encore un retour de 1
        }
    }
}

// Enregistrer l'extension combinée dans TurboWarp
Scratch.extensions.register(new CombinedExtension());
`

**Comment by Eaielectronic** [comment]
lass CombinedExtension {
    constructor() {
        this.previousTime = null;  // Stocke l'heure de la dernière frame
        this.count = 0;            // Compte le nombre de fois où on a retourné 1
    }

    getInfo() {
        return {
            id: 'combinedextension',
            name: 'Combined Delta Time & Return 1',
            blocks: [
                {
                    opcode: 'getDeltaTime',
                    blockType: Scratch.BlockType.REPORTER,
                    text: 'delta time'
                },
                {
                    opcode: 'getFixedDelta',
                    blockType: Scratch.BlockType.REPORTER,
                    text: 'fixed delta (target 1/sec)'
                },
                {
                    opcode: 'returnOneMultipleTimes',
                    blockType: Scratch.BlockType.REPORTER,
                    text: 'return 30 times 1 every second'
                },
                {
                    opcode: 'getIncrementForDuration',
                    blockType: Scratch.BlockType.REPORTER,
                    text: 'increment to reach [TARGET] in [DURATION] sec',
                    arguments: {
                        TARGET: {
                            type: Scratch.ArgumentType.NUMBER,
                            defaultValue: 1
                        },
                        DURATION: {
                            type: Scratch.ArgumentType.NUMBER,
                            defaultValue: 3
                        }
                    }
                }
            ]
        };
    }

    // Méthode pour obtenir le delta time
    getDeltaTime() {
        const currentTime = performance.now();

        if (this.previousTime === null) {
            this.previousTime = currentTime;
            return 0;
        }

        let deltaTime = (currentTime - this.previousTime) / 1000; // Convertir en secondes
        this.previousTime = currentTime;

        if (deltaTime > 1/30) {
            deltaTime = 1/30;  // Limiter à 30 FPS max pour éviter les pics
        }

        return deltaTime;
    }

    // Méthode pour obtenir un delta fixe
    getFixedDelta() {
        const deltaTime = this.getDeltaTime();
        return deltaTime * 100;  // Calibrer pour que "100 * delta time" = 1 unité ajoutée par seconde
    }

    // Méthode pour obtenir un incrément en fonction de la durée pour atteindre une cible
    getIncrementForDuration(args) {
        const deltaTime = this.getDeltaTime();
        const targetValue = args.TARGET;   // Valeur cible que vous voulez atteindre
        const duration = args.DURATION;    // Durée en secondes pour atteindre la valeur

        // Chaque seconde, on ajoute une fraction pour atteindre la valeur cible dans le temps imparti
        return (targetValue / duration) * deltaTime;
    }

    // Méthode pour retourner 1 plusieurs fois par seconde
    returnOneMultipleTimes() {
        const deltaTime = this.getDeltaTime();
        const returnsPerSecond = 30;  // Retourner 30 fois 1 par seconde
        const returns = deltaTime * returnsPerSecond;  // Combien de fois retourner 1 durant deltaTime

        this.count += returns;

        if (this.count >= 1) {
            this.count -= 1;
            return 1;
        } else {
            return 0;  // Retourner 0 si le temps écoulé ne justifie pas encore un retour de 1
        }
    }
}

// Enregistrer l'extension combinée dans TurboWarp
Scratch.extensions.register(new CombinedExtension());

**Reply by Thebloxers998** [reply]
Noice

**Comment by Thebloxers998** [comment]
You messed up the code though

**Reply by Eaielectronic** [reply]
sorry for the messy code, I created a repository with interesting extensions: one to create a login form, one to detect fingers on ALL mobile devices (sensing + does not work on iOS apparently) and another with a block that returns thirty times 1 in a second (to create games that work with different fps and it is not the fps block of the extension.) PS: sometimes the touch extension does not work in the first script and when you display it on the screen it does not work sometimes but otherwise it works perfectly for the rest and you just have to delete the first script if it does not work and reimport it. the link to the repository is here: https://github.com/Eaielectronic/Extension-TurboWarp/tree/main.

**Comment by Eaielectronic** [comment]
sorry for the messy code, I created a repository with interesting extensions: one to create a login form, one to detect fingers on ALL mobile devices (sensing + does not work on iOS apparently) and another with a block that returns thirty times 1 in a second (to create games that work with different fps and it is not the fps block of the extension.) PS: sometimes the touch extension does not work in the first script and when you display it on the screen it does not work sometimes but otherwise it works perfectly for the rest and you just have to delete the first script if it does not work and reimport it. the link to the repository is here: https://github.com/Eaielectronic/Extension-TurboWarp/tree/main.

**Reply by Thebloxers998** [reply]
I'm guessing login page will be a great extension

**Comment by Eaielectronic** [comment]
for scrolling bubble, it creates text that appears letter by letter, you can put words in bold, in color or in italics with the following keys (json) :  
Une fois touché, les {"type":"gras","color":"#e2c509","mot":"drapeaux"} vous permettent de {"type":"gras","color":"#00bfff","mot":"réapparaître"} au dernier ouvert !

Pour bouger, appuis sur les {"type":"gras","color":"#5bd691","mot":"flêches"}.

**Comment by Thebloxers998** [comment]
I also have a login signin extension, its called TurboAuth (TurboWarpAuthenticator)

**Reply by Thebloxers998** [reply]
I haven't finished yet though

**Reply by Thebloxers998** [reply]
You can rate it if you want
